src/Util.cpp

```diff
@@ -411,62 +457,145 @@ std::vector<double> ale::getDistortionCoeffs(json isd) {
   return coefficients;
 }
 
-std::vector<double> ale::getSunPositions(json isd) {
-  std::vector<double> positions;
+std::vector<Vec3d> getJsonVec3dArray(json obj) {
+  std::vector<Vec3d> positions;
   try {
-    json jayson = isd.at("sun_position");
-    for (auto &location : jayson.at("positions")) {
-      positions.push_back(location[0].get<double>());
-      positions.push_back(location[1].get<double>());
-      positions.push_back(location[2].get<double>());
+    for (auto &location : obj) {
+      Vec3d vec(location[0].get<double>(),location[1].get<double>(), location[2].get<double>() );
+      positions.push_back(vec);
     }
   } catch (...) {
-    throw std::runtime_error("Could not parse the sun positions.");
+    throw std::runtime_error("Could not parse the 3D vector array.");
   }
   return positions;
 }
 
-std::vector<double> ale::getSensorPositions(json isd) {
-  std::vector<double> positions;
+
+std::vector<Rotation> getJsonQuatArray(json obj) {
+  std::vector<Rotation> quats;
   try {
-    json jayson = isd.at("sensor_position");
-    for (auto &location : jayson.at("positions")) {
-      positions.push_back(location[0].get<double>());
-      positions.push_back(location[1].get<double>());
-      positions.push_back(location[2].get<double>());
+    for (auto &location : obj) {
+      Rotation vec(location[0].get<double>(),location[1].get<double>(), location[2].get<double>(), location[2].get<double>() );
+      quats.push_back(vec);
     }
   } catch (...) {
-    throw std::runtime_error("Could not parse the sensor positions.");
+    throw std::runtime_error("Could not parse the quaternion json object.");
   }
-  return positions;
+  return quats;
 }
 
-std::vector<double> ale::getSensorVelocities(json isd) {
-  std::vector<double> velocities;
+
+States getInstrumentPosition(json isd) {
   try {
-    json jayson = isd.at("sensor_position");
-    for (auto &velocity : jayson.at("velocities")) {
-      velocities.push_back(velocity[0].get<double>());
-      velocities.push_back(velocity[1].get<double>());
-      velocities.push_back(velocity[2].get<double>());
+    json ipos = isd.at("instrument_position");
+    std::vector<Vec3d> positions = getJsonVec3dArray(ipos.at("positions"));
+    std::vector<double> times = getJsonArray<double>(ipos.at("ephemeris_times")); 
+    int refFrame = ipos.at("reference_frame").get<int>();
+    
+    bool hasVelocities = ipos.find("velocities") != ipos.end();
+    if (hasVelocities) {
+      std::vector<Vec3d> velocities = getJsonVec3dArray(ipos.at("velocities")); 
+      States states(times, positions, velocities, refFrame);  
+      return states;
```

Comment:
Could also just do return `States(times, positions, velocities, refFrame);` to save a line.